File: src/main/kotlin/top/fpsmaster/web/network/packets/GuiLoadPackets.kt
package top.fpsmaster.web.network.packets

import top.fpsmaster.web.network.packet.ClientboundPacket
import top.fpsmaster.web.network.packet.ServerboundPacket
import top.fpsmaster.web.network.serializer.PacketBuffer

/**
 * GUI加载事件数据包
 *
 * 从Minecraft发送到UI，触发UI的加载动画
 */
class GuiLoadEventPacket : ClientboundPacket() {
    var eventType: String = "open"  // open, close, refresh
    var timestamp: Long = System.currentTimeMillis()
    var extraData: String? = null   // 额外数据

    override fun write(buffer: PacketBuffer) {
        buffer.writeString(eventType)
        buffer.writeLong(timestamp)
        buffer.writeString(extraData)
    }

    override fun read(buffer: PacketBuffer) {
        eventType = buffer.readString() ?: "open"
        timestamp = buffer.readLong()
        extraData = buffer.readString()
    }

    override fun toString(): String {
        return "GuiLoadEventPacket(event='$eventType', timestamp=$timestamp, data=$extraData)"
    }
}

/**
 * GUI加载确认数据包
 *
 * UI确认收到加载事件，Minecraft收到后开始渲染
 */
class GuiLoadAckPacket : ServerboundPacket() {
    var success: Boolean = false
    var message: String = ""
    var timestamp: Long = System.currentTimeMillis()

    override fun write(buffer: PacketBuffer) {
        buffer.writeBoolean(success)
        buffer.writeString(message)
        buffer.writeLong(timestamp)
    }

    override fun read(buffer: PacketBuffer) {
        success = buffer.readBoolean()
        message = buffer.readString() ?: ""
        timestamp = buffer.readLong()
    }

    override fun toString(): String {
        return "GuiLoadAckPacket(success=$success, message='$message', timestamp=$timestamp)"
    }
}

File: ui/network/packets/GuiLoadPackets.ts
import { Packet } from '../Packet';
import { PacketBuffer } from '../PacketBuffer';

export class GuiLoadEventPacket implements Packet {
  packetId = 9;
  eventType: string = "open";
  timestamp: number = 0;
  extraData: string | null = null;

  write(buffer: PacketBuffer): void {
    buffer.writeString(this.eventType);
    buffer.writeLong(this.timestamp);
    buffer.writeString(this.extraData);
  }

  read(buffer: PacketBuffer): void {
    this.eventType = buffer.readString() || "open";
    this.timestamp = buffer.readLong();
    this.extraData = buffer.readString();
  }
}

export class GuiLoadAckPacket implements Packet {
  packetId = 10;
  success: boolean = false;
  message: string = "";
  timestamp: number = 0;

  constructor(success: boolean = false, message: string = "", timestamp: number = Date.now()) {
    this.success = success;
    this.message = message;
    this.timestamp = timestamp;
  }

  write(buffer: PacketBuffer): void {
    buffer.writeBoolean(this.success);
    buffer.writeString(this.message);
    buffer.writeLong(this.timestamp);
  }

  read(buffer: PacketBuffer): void {
    this.success = buffer.readBoolean();
    this.message = buffer.readString() || "";
    this.timestamp = buffer.readLong();
  }
}
